test/Bucket/bucket_test.jl

```diff
@@ -0,0 +1,305 @@
+        parent(
+            p.bucket.R_n .-
+            -(290.0 * 0.8 + 10.0 - 5.67e-8 * 255.69458073555182^4.0),
+        ),
+    ) < eps(FT)
+
+
+end
+
+@testset "Moisture Conservation" begin
+    # Ensure no heat fluxes (no evaporation too, so no LHF)
+    # and check that runoff+soil water balance precip.
+    "Radiation"
+    SW_d = (t) -> eltype(t)(0.0)
+    LW_d = (t) -> eltype(t)(280.0^4.0 * 5.67e-8) # zeros out LW radiation
+    bucket_rad = PrescribedRadiation(FT, SW_d, LW_d)
+    # Atmosphere with no precipitation
+    precip = (t) -> eltype(t)(1e-2)
+    T_atmos = (t) -> eltype(t)(298.0)
+    u_atmos = (t) -> eltype(t)(0.0) # no turbulent fluxes
+    q_atmos = (t) -> eltype(t)(0.0) # no atmos water
+    h_atmos = FT(30)
+    ρ_atmos = (t) -> eltype(t)(1.13)
+    ρ_sfc = FT(1.15)
+    bucket_atmos = PrescribedAtmosphere(
+        precip,
+        T_atmos,
+        u_atmos,
+        q_atmos,
+        ρ_atmos,
+        h_atmos,
+        ρ_sfc,
+    )
+
+    model = BucketModel(
+        parameters = bucket_parameters,
+        domain = domain,
+        atmosphere = bucket_atmos,
+        radiation = bucket_rad,
+    )
+
+    # Initial conditions with no moisture
+    Y, p, coords = initialize(model)
+    Y.bucket.T_sfc .= 280.0
+    Y.bucket.W .= 0.14 # no moisture
+    Y.bucket.Ws .= 0.0 # no runoff
+    Y.bucket.S .= 0.0
+    ode_function! = make_ode_function(model)
+
+    t0 = 0.0
+    tf = 10.0
+    prob = ODEProblem(ode_function!, Y, (t0, tf), p)
+    Δt = 1.0
+    # need a callback to get `p`
+    sol = solve(prob, RK4(); dt = Δt, saveat = 0:Δt:tf)
+    # net soil water
+    land_water = unique(parent(sol.u[end].bucket.W .+ sol.u[end].bucket.Ws))[1]
+    dland_water = land_water - 0.14 # 0.14 is initial water
+    datmos_water = 1e-2 * sol.t[end]
+    @test datmos_water ≈ dland_water
+end
+
+@testset "Energy + Moisture Conservation" begin
+    "Radiation"
+    SW_d = (t) -> eltype(t)(10.0)
+    LW_d = (t) -> eltype(t)(300.0)
+    bucket_rad = PrescribedRadiation(FT, SW_d, LW_d)
+    "Atmos"
+    precip = (t) -> eltype(t)(1e-6)
+    T_atmos = (t) -> eltype(t)(298.0)
+    u_atmos = (t) -> eltype(t)(4.0)
+    q_atmos = (t) -> eltype(t)(0.01)
+    h_atmos = FT(30)
+    ρ_atmos = (t) -> eltype(t)(1.13)
+    ρ_sfc = FT(1.15)
+    bucket_atmos = PrescribedAtmosphere(
+        precip,
+        T_atmos,
+        u_atmos,
+        q_atmos,
+        ρ_atmos,
+        h_atmos,
+        ρ_sfc,
+    )
+
+    model = BucketModel(
+        parameters = bucket_parameters,
+        domain = domain,
+        atmosphere = bucket_atmos,
+        radiation = bucket_rad,
+    )
+
+    # Initial conditions with no moisture
+    Y, p, coords = initialize(model)
+    Y.bucket.T_sfc .= 280.0
+    Y.bucket.W .= 0.149
+    Y.bucket.Ws .= 0.0
+    Y.bucket.S .= 0.0
+    ode_function! = make_ode_function(model)
+
+    t0 = 0.0
+    tf = 10000.0
+    prob = ODEProblem(ode_function!, Y, (t0, tf), p)
+    Δt = 100.0
+    # need a callback to get `p`
```

Comment:
Would using DiffEq's `init` and `step!` work for the time stepping here? `p` (as well as other useful info) could then be accessed via the integrator.